src/dbnode/persist/fs/commitlog/commit_log.go

```diff
@@ -506,69 +538,154 @@ func (l *commitLog) write() {
+		// scenarios described above.
+		if len(writer.pendingFlushFns) == 0 {
+			l.metrics.flushDone.Inc(1)
+			return
+		}
 
-	for i := range l.pendingFlushFns {
-		l.pendingFlushFns[i](callbackResult{
-			eventType: flushEventType,
-			err:       err,
-		})
-		l.pendingFlushFns[i] = nil
+		for i := range writer.pendingFlushFns {
+			writer.pendingFlushFns[i](callbackResult{
+				eventType: flushEventType,
+				err:       err,
+			})
+			writer.pendingFlushFns[i] = nil
+		}
+		writer.pendingFlushFns = writer.pendingFlushFns[:0]
+		l.metrics.flushDone.Inc(1)
 	}
-	l.pendingFlushFns = l.pendingFlushFns[:0]
-	l.metrics.flushDone.Inc(1)
 }
 
 // writerState lock must be held for the duration of this function call.
-func (l *commitLog) openWriter() (persist.CommitLogFile, error) {
-	if l.writerState.writer != nil {
-		if err := l.writerState.writer.Close(); err != nil {
-			l.metrics.closeErrors.Inc(1)
-			l.log.Error("failed to close commit log", zap.Error(err))
-
-			// If we failed to close then create a new commit log writer
-			l.writerState.writer = nil
+func (l *commitLog) openWriters() (persist.CommitLogFiles, error) {
+	// Ensure that the previous asynchronous reset of the secondary writer (if any)
+	// has completed before attempting to start a new one and/or modify the writerState
+	// in any way.
+	l.waitForSecondaryWriterAsyncResetComplete()
+
+	if l.writerState.primaryWriter.writer == nil || l.writerState.secondaryWriter.writer == nil {
+		// If either of the commitlog writers is nil then open both of them synchronously. Under
+		// normal circumstances this will only occur when the commitlog is first opened.
+		primaryWriterFlushFn := l.newOnFlushFn(&l.writerState.primaryWriter)
+		secondaryWriterFlushFn := l.newOnFlushFn(&l.writerState.secondaryWriter)
+		l.writerState.primaryWriter.writer = l.newCommitLogWriterFn(primaryWriterFlushFn, l.opts)
+		l.writerState.secondaryWriter.writer = l.newCommitLogWriterFn(secondaryWriterFlushFn, l.opts)
+
+		primaryFile, err := l.writerState.primaryWriter.writer.Open()
+		if err != nil {
+			return nil, err
 		}
-	}
 
-	if l.writerState.writer == nil {
-		l.writerState.writer = l.newCommitLogWriterFn(l.onFlush, l.opts)
-	}
+		secondaryFile, err := l.writerState.secondaryWriter.writer.Open()
+		if err != nil {
+			return nil, err
+		}
 
-	file, err := l.writerState.writer.Open()
-	if err != nil {
-		return persist.CommitLogFile{}, err
+		l.writerState.activeFiles = persist.CommitLogFiles{primaryFile, secondaryFile}
+		return l.writerState.activeFiles, nil
 	}
 
-	l.writerState.activeFile = &file
+	// Swap the primary and secondary writers so that the secondary becomes primary and vice versa.
+	// This consumes the standby secondary writer, but a new one will be prepared asynchronously by
+	// resetting the formerly primary writer.
+	prevPrimary := l.writerState.primaryWriter
+	l.writerState.primaryWriter = l.writerState.secondaryWriter
+	l.writerState.secondaryWriter = prevPrimary
```

Comment:
Maybe:
`l.writerState.primaryWriter, l.writerState.secondaryWriter = l.writerState.secondaryWriter, l.writerState.primaryWriter`

Reply:
yeah I just broke it up because its an extremely long line

Reply:
`l.writerState.primary, l.writerState.secondary = l.writerState.secondary, l.writerState.primary` is a little shorter if you trim `primaryWriter` -> `primary` and same with secondary.